### ISSUE TITLE
backonline not working for me

### ISSUE THREAD

**Issue by hughperkins** [Element-Research/rnn#98]
backonline isnt working for me.  even in the simplest case, without any lstm at all.  thoughts?  For simplicity, I have a single batch, of length 2, and just repeat the same batch/sequence over and over, exact same batch/sequence.  So, should overtrain pretty quickly.  But it doesnt, at least not, for backwardsonline.

If I train without using any rnn code, ie using https://github.com/hughperkins/char-rnn-er/blob/debugging/train.lua#L303-L327 : 

```
    for s=1,seqLength do
      net:forget()
      net:zeroGradParameters()
      local batchOffset = (epochOffset + (it - 1) * seqLength + (s - 1) + 1 - 1) % input_len + 1
      local targetOffset = (batchOffset + 1 - 1) % input_len + 1

      populateBatchInput(batchOffset, debugState, inputStriped, batchInput)
      local batchTarget = makeBatchTarget(targetOffset, debugState, inputStriped)

      local batchOutput = net:forward(batchInput)
      local batchLoss = crit:forward(batchOutput, batchTarget)
      local batchGradOutput = crit:backward(batchOutput, batchTarget)
      net:backward(batchInput, batchGradOutput)

      net:updateParameters(learningRate)
    end
```
... then everything works just fine.  output looks like:
```
batchOutput:exp()	 0.0104  0.9896
batchLoss	-0.98956722021103	
batchOutput:exp()	 0.9896  0.0104
[torch.CudaTensor of size 1x2]
batchLoss	-0.98963707685471	
```
In other words:
- for s=1, the second output neuron is high, and the first is low
- and the loss is near to -1
- for s=2, the first output neuron is high, and the second is low
- and the loss is near to -1
... which is what we would want.  (So, basically, I'm training with input='F', target='i', and input='i', output='F').

Now, if I use the exact same training data, ie 'F' => 'i', and 'i' => 'F', but training using backwardsonline, as a sequence of 2, using https://github.com/hughperkins/char-rnn-er/blob/debugging/train.lua#L264-L299

```
    batchInputs = {}
    batchOutputs = {}
    batchOffsets = {}

    net:forget()
    net:zeroGradParameters()
    net:backwardOnline()
    for s=1,seqLength do
      local batchOffset = (epochOffset + (it - 1) * seqLength + (s - 1) + 1 - 1) % input_len + 1
      batchOffsets[s] = batchOffset
      populateBatchInput(batchOffset, debugState, inputStriped, batchInput)
      local batchOutput = net:forward(batchInput)
      batchInputs[s] = batchInput
      batchOutputs[s] = batchOutput
    end

    for s=seqLength,1,-1 do
      local targetOffset = (batchOffsets[s] + 1 - 1) % input_len + 1
      local batchTarget = makeBatchTarget(targetOffset, debugState, inputStriped)
      local batchLoss = crit:forward(batchOutputs[s], batchTarget)
      local batchGradOutput = crit:backward(batchOutputs[s], batchTarget)
      net:backward(batchInputs[s], batchGradOutput)
    end
    net:updateParameters(learningRate)
```
... then the loss totally misbehaves, and the outputs are uniform across all output neurons:
```
batchOutput:exp()	 0.3777  0.6223
batchOutput:exp()	 0.5000  0.5000
batchLoss	-0.50000023841858	
batchLoss	-0.49999976158142	
```
- for s=1, the second neuron is somewhat larger than the first
- for 2=2, both neurons give identical outputs
- in both cases, the loss is -0.5, rather than -1.0

Learning rate is same in both cases, and the network is really simply, not even an LSTM in sight (to keep it simple; I get the same effect with an LSTM in too, but trying to simplify the program, until the root cause becomes obvoius):

```
net	nn.Sequential {
  [input -> (1) -> (2) -> output]
  (1): nn.Linear(2 -> 2)
  (2): nn.LogSoftMax
}
```
Thoughts?  Something obvoiusly dumb/invalid I'm doing in the `backwardsonline` case?

**Comment by hughperkins** [comment]
Oh, I just realized, it's because I'm reusing `batchInput` across the different time values in the sequence... I'll fix that :-P

**Comment by hughperkins** [comment]
(and output from each `:forward()` I guess should be copied/cloned too actually I guess?)

**Comment by hughperkins** [comment]
Well... I still have an issue now.  But it's slightly different.  But similar.  Basically, if I have the same sequence of 4 different inputs, which are basically: 1=>2, 2=>3, 3=>4, 4=>1, then the outputs, and losses are:

```
batchOutput:exp()	 0.0000  1.0000  0.0000  0.0000
[torch.CudaTensor of size 1x4]

batchOutput:exp()	 0.0000  0.0000  1.0000  0.0000
[torch.CudaTensor of size 1x4]

batchOutput:exp()	 0.0000  0.0000  0.0000  1.0000
[torch.CudaTensor of size 1x4]

batchOutput:exp()	 0.3993  0.2002  0.2002  0.2002
[torch.CudaTensor of size 1x4]

batchLoss	0.91801500320435	
batchLoss	0	
batchLoss	0	
batchLoss	0	
```
So, bearing in mind that the outputs are in forward order, and the losses are in backwards order, in this print out, basically all the outputs and losses are correct, except for the final one in the sequence.

If I just use normal non-sequential backprop, one forwards/backwards for each pair of input/output, then the output is correct.  Although strangely the loss goes to zero *signicantly* slower than with backwardsonline.  With identical network and learning rate:

```
batchOutput:exp()	 0.0164  0.9522  0.0133  0.0182
[torch.CudaTensor of size 1x4]
batchLoss	-0.95219099521637	

batchOutput:exp()	 0.0176  0.0152  0.9526  0.0146
[torch.CudaTensor of size 1x4]
batchLoss	-0.95260453224182	

batchOutput:exp()	 0.0168  0.0145  0.0173  0.9514
[torch.CudaTensor of size 1x4]
batchLoss	-0.95141333341599	

batchOutput:exp()	 0.9503  0.0175  0.0162  0.0161
[torch.CudaTensor of size 1x4]
batchLoss	-0.95030152797699	
```
(here, both the output and the loss are in forwards sequential order)

Relevant code is:

backsonline, https://github.com/hughperkins/char-rnn-er/blob/debugging/train.lua#L306-L342 :
```
    net:forget()
    net:zeroGradParameters()
    net:backwardOnline()
    for s=1,seqLength do
      batchOffsets[s] = lua_modulus(epochOffset + (it - 1) * seqLength + (s - 1) + 1, input_len)

      timer_update(timer, 'forward setup')
      populateBatchInput(batchOffsets[s], debugState, inputStriped, batchInputs[s])

      timer_update(timer, 'forward run')
      local batchOutput = net:forward(batchInputs[s])
--      batchInputs[s] = batchInput
      batchOutputs[s]:copy(batchOutput)

      doOutputDebug(debugState, batchOutput)
    end

    for s=seqLength,1,-1 do
      timer_update(timer, 'backward setup')
      local targetOffset = lua_modulus(batchOffsets[s] + 1, input_len)
      local batchTarget = makeBatchTarget(targetOffset, debugState, inputStriped)

      timer_update(timer, 'backward run')
      local batchLoss = crit:forward(batchOutputs[s], batchTarget)
      local batchGradOutput = crit:backward(batchOutputs[s], batchTarget)
      net:backward(batchInputs[s], batchGradOutput)

--      if debugState.printOutput then
--        print('batchInputs[s]', batchInputs[s])
--        print('batchTarget', batchTarget)
--        print('batchGradOutput', batchGradOutput)
--      end

      seqLoss = seqLoss + batchLoss
      doBackwardDebug(debugState, batchTarget, batchLoss, batchGradOutput)
    end
    net:updateParameters(learningRate)
```

no sequential, https://github.com/hughperkins/char-rnn-er/blob/debugging/train.lua#L346-L373 :
```
    for s=1,seqLength do
      net:forget()
      net:zeroGradParameters()
      local batchOffset = lua_modulus(epochOffset + (it - 1) * seqLength + (s - 1) + 1, input_len)
      local targetOffset = lua_modulus(batchOffset + 1, input_len)

      timer_update(timer, 'forward setup')
      populateBatchInput(batchOffset, debugState, inputStriped, batchInputs[s])

      timer_update(timer, 'backward setup')
      local batchTarget = makeBatchTarget(targetOffset, debugState, inputStriped)

      timer_update(timer, 'forward run')
      local batchOutput = net:forward(batchInputs[s])
      doOutputDebug(debugState, batchOutput)

      timer_update(timer, 'backward run')
      local batchLoss = crit:forward(batchOutput, batchTarget)
      local batchGradOutput = crit:backward(batchOutput, batchTarget)
      net:backward(batchInputs[s], batchGradOutput)
      if debugState.printOutput then
        
      end
      net:updateParameters(learningRate)

      seqLoss = seqLoss + batchLoss
      doBackwardDebug(debugState, batchTarget, batchLoss, batchGradOutput)
    end
```

**Comment by hughperkins** [comment]
Update: note that backwardThroughTime works correctly, and gives the same rather slow decay of the loss as non-rnn updates:

```
batchOutput:exp()	 0.0042  0.9869  0.0044  0.0046
[torch.CudaTensor of size 1x4]

batchLoss	0.01317310333252	
batchOutput:exp()	 0.0044  0.0043  0.9869  0.0044
[torch.CudaTensor of size 1x4]

batchLoss	0.013160705566406	
batchOutput:exp()	 0.0045  0.0045  0.0042  0.9868
[torch.CudaTensor of size 1x4]

batchLoss	0.01329231262207	
batchOutput:exp()	 0.9869  0.0043  0.0045  0.0043
[torch.CudaTensor of size 1x4]

batchLoss	0.013157844543457	
params:narrow(1,1,seqLength)	-4.0845 -3.8753 -3.6628  1.3514
[torch.CudaTensor of size 1x4]
```

...but `backwardThroughTime()` cannot be used with a network with more than one LSTM layer, per my understanding?

**Comment by hughperkins** [comment]
Ah, apparenlty it's because I need to wrap a `Recursor` around my network.  Figured it out after writing some notes/documentation at https://github.com/hughperkins/char-rnn-er/blob/master/doc/rnn_summary.md . Updated network to be wrapped in Recursor in https://github.com/hughperkins/char-rnn-er/commit/687a22daac4c222e5bf40f65660b39dfaa68afae

**Comment by nicholas-leonard** [comment]
@hughperkins Glad you resolved this. Love the rnn_summary.md notes. Sorry for all the confusion this has caused you.

**Comment by hughperkins** [comment]
Thanks!  Gradually working my way through, but looking promising so far :-)
